### PR TITLE
Refactor image tree for API usage

### PR DESCRIPTION
**Pull request by saschagrunert**
Refers to https://github.com/containers/libpod/issues/2791#issuecomment-581956222

### CODE REVIEW

**Comment by rhatdan** [comment]
@baude PTAL

**Comment by vrothberg** [comment]
Linter isn't happy yet. Will review now :+1: 
```
[+0036s] cmd/podman/tree.go:55:13: SA1006: printf-style function with dynamic first argument and no further arguments should use print-style function instead (staticcheck)
[+0036s] 	fmt.Printf(tree)
```

File: libpod/image/tree.go [hunk 1]
```diff
@@ -0,0 +1,137 @@
...
+)
+
+type tree struct {
+	img       *Image
+	imageInfo *InfoImage
+	layerInfo map[string]*LayerInfo
+	sb        *strings.Builder
+}
+
+// Tree ...
```
Comment (vrothberg) [inline libpod/image/tree.go]: Please add a verbose comment.

suggestion: `s/NewTree/Tree`

**Comment by saschagrunert** [comment]
> Linter isn't happy yet. Will review now 

Thanks! Please keep in mind that the varlink API does not work yet :)

**Comment by vrothberg** [comment]
@baude @jwhonce PTAL

**Comment by saschagrunert** [comment]
Varlink API works now, ready for review.

File: libpod/image/tree.go [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+type tree struct {
+	img       *Image
+	imageInfo *InfoImage
+	layerInfo map[string]*LayerInfo
+	sb        *strings.Builder
+}
+
+// NewTree creates an image tree string representation for displaying it to the
+// user.
+func (i *Image) NewTree(whatRequires bool) (string, error) {
```
Comment (vrothberg) [inline libpod/image/tree.go]: I still don't like `NewTree` as it's not a constructor for a `Tree` object but it generates a string. I'd much prefer `Tree` or `GenerateTree`.

**Comment by vrothberg** [comment]
> Can you also extend the swagger docs for the image endpoint? -> https://github.com/containers/libpod/blob/master/pkg/api/server/register_images.go#L581

Still open.

File: pkg/api/handlers/libpod/images.go [hunk 3]
```diff
@@ -38,17 +38,22 @@ func ImageExists(w http.ResponseWriter, r *http.Request) {
...
+	runtime := r.Context().Value("runtime").(*libpod.Runtime)
+	name := utils.GetName(r)
+
+	img, err := runtime.ImageRuntime().NewFromLocal(name)
+	if err != nil {
+		utils.Error(w, "Something went wrong.", http.StatusNotFound, errors.Wrapf(err, "Failed to find image %s", name))
+		return
+	}
+
+	tree, err := img.NewTree(true)
```
Comment (vrothberg) [inline pkg/api/handlers/libpod/images.go]: We need a whatrequires paramater here as well.

File: libpod/image/tree.go (hunk 2 above)
Comment (saschagrunert) [reply libpod/image/tree.go]: Alright, renamed it to `GenerateTree`

File: pkg/api/handlers/libpod/images.go (hunk 3 above)
Comment (saschagrunert) [reply pkg/api/handlers/libpod/images.go]: Added the parameter as well.

**Comment by vrothberg** [comment]
@saschagrunert, can you rebase? This will make local tests easier as we'd fetch `podman system service`.

File: pkg/api/server/register_images.go [hunk 4]
```diff
@@ -578,6 +578,29 @@ func (s *APIServer) registerImagesHandlers(r *mux.Router) error {
...
+	//    description: the name or ID of the container
+	//  - in: query
+	//    name: whatrequires
+	//    type: bool
+	//    description: show all child images and layers of the specified image
+	// produces:
+	// - application/json
+	// responses:
+	//   200:
+	//     $ref: "#/response/LibpodImageTreeResponse"
```
Comment (vrothberg) [inline pkg/api/server/register_images.go]: Need a...
```
//   404:
//     $ref: '#/responses/NoSuchImage'
```

File: pkg/api/handlers/libpod/images.go [hunk 5]
```diff
@@ -38,17 +38,23 @@ func ImageExists(w http.ResponseWriter, r *http.Request) {
...
+	runtime := r.Context().Value("runtime").(*libpod.Runtime)
+	name := utils.GetName(r)
+
+	img, err := runtime.ImageRuntime().NewFromLocal(name)
+	if err != nil {
+		utils.Error(w, "Something went wrong.", http.StatusNotFound, errors.Wrapf(err, "Failed to find image %s", name))
+		return
+	}
+
+	whatRequires := r.Context().Value("whatrequires").(bool)
```
Comment (vrothberg) [inline pkg/api/handlers/libpod/images.go]: 2020/02/14 16:44:29 http: panic serving @: interface conversion: interface {} is nil, not bool
goroutine 229 [running]:
net/http.(*conn).serve.func1(0xc0006ca3c0)
        /usr/lib/golang/src/net/http/server.go:1767 +0x139
panic(0x18b46a0, 0xc00021b5f0)
        /usr/lib/golang/src/runtime/panic.go:679 +0x1b2
github.com/containers/libpod/pkg/api/handlers/libpod.ImageTree(0x1e48320, 0xc0002367e0, 0xc0003ac900)
        pkg/api/handlers/libpod/images.go:58 +0x405
github.com/containers/libpod/pkg/api/server.APIHandler.func1(0x1e48320, 0xc0002367e0, 0xc0003ac600)
        pkg/api/server/handler_api.go:24 +0x47e
net/http.HandlerFunc.ServeHTTP(0xc00041cec0, 0x1e48320, 0xc0002367e0, 0xc0003ac600)
        /usr/lib/golang/src/net/http/server.go:2007 +0x44
github.com/gorilla/mux.(*Router).ServeHTTP(0xc00025ecc0, 0x1e48320, 0xc0002367e0, 0xc0003ac400)
        vendor/github.com/gorilla/mux/mux.go:210 +0xe2
net/http.serverHandler.ServeHTTP(0xc0005357a0, 0x1e48320, 0xc0002367e0, 0xc0003ac400)
        /usr/lib/golang/src/net/http/server.go:2802 +0xa4
net/http.(*conn).serve(0xc0006ca3c0, 0x1e51560, 0xc000533c40)
        /usr/lib/golang/src/net/http/server.go:1890 +0x875
created by net/http.(*Server).Serve
        /usr/lib/golang/src/net/http/server.go:2928 +0x384

Comment (vrothberg) [reply pkg/api/handlers/libpod/images.go]:
```golang
query := struct {
        Filters map[string][]string `schema:"filters"`
}{
        // override any golang type defaults
}

if err := decoder.Decode(&query, r.URL.Query()); err != nil {
        utils.Error(w, http.StatusText(http.StatusBadRequest), http.StatusBadRequest,
                errors.Wrapf(err, "Failed to parse parameters for %s", r.URL.String()))
        return
}
```

Is what most endpoints do.

File: pkg/api/server/register_images.go (hunk 4 above)
Comment (saschagrunert) [reply pkg/api/server/register_images.go]: Alright, added that as well. :+1:

File: pkg/api/handlers/libpod/images.go (hunk 5 above)
Comment (saschagrunert) [reply pkg/api/handlers/libpod/images.go]: Ah alright thank you for the hint. Changed as suggested :+1:

**Comment by saschagrunert** [comment]
> @saschagrunert, can you rebase? This will make local tests easier as we'd fetch `podman system service`.

Rebased on top of the latest master branch and addressed your review comments.

**Comment by saschagrunert** [comment]
Docs seem to be green now as well :)

File: pkg/api/handlers/libpod/images.go [hunk 6]
```diff
@@ -46,17 +46,34 @@ func ImageExists(w http.ResponseWriter, r *http.Request) {
...
-	// }
-	//	it is not clear to me how to deal with this given all the processing of the image
-	// is in main.  we need to discuss how that really should be and return something useful.
-	handlers.UnsupportedHandler(w, r)
+	runtime := r.Context().Value("runtime").(*libpod.Runtime)
+	name := utils.GetName(r)
+
+	img, err := runtime.ImageRuntime().NewFromLocal(name)
+	if err != nil {
+		utils.Error(w, "Something went wrong.", http.StatusNotFound, errors.Wrapf(err, "Failed to find image %s", name))
```
Comment (TomSweeneyRedHat) [inline pkg/api/handlers/libpod/images.go]: I hate "Something went wrong. "  Can we change it to just "Error:"

Comment (baude) [reply pkg/api/handlers/libpod/images.go]: no we cannot

**Comment by baude** [comment]
it LGTM but i want @mheon to peek at this one

File: libpod/image/tree.go [hunk 7]
```diff
@@ -0,0 +1,138 @@
+package image
```
Comment (mheon) [inline libpod/image/tree.go]: Can we move this to somewhere in pkg/? It doesn't seem much like an API call, given it only returns a string.

Comment (vrothberg) [reply libpod/image/tree.go]: I suggest to wait for @mtrmac's new package. We may be able to integrate it there.

Comment (vrothberg) [reply libpod/image/tree.go]: By "wait" I mean we could merge it as is now and let Miloslav pick it up in his new package.

Comment (mheon) [reply libpod/image/tree.go]: Hm. It's just libpod/image, not libpod itself, so sure

Comment (mtrmac) [reply libpod/image/tree.go]: Definitely don’t wait for me, I’m still mostly in learning phase, and I can adjust for whatever happens here.

(Jumping into what I’m sure was a well-considered decision, I’d instinctively prefer to primarily have a well-typed API that returns the structure, and keep the text formatter in the CLI (allowing, at least in principle, a nice GUI to be built on top, for example). OTOH, I guess that exposing the complete tree structure over Varlink/whatever would be a lot of work and pain — and for no benefit at least now that the only consumer is the CLI.)

**Comment by mheon** [comment]
LGTM

File: libpod/image/tree.go (hunk 7 above)
Comment (mheon) [reply libpod/image/tree.go]: That's my big hold-up as well - I'd love to return a data structure that represents the tree, with a `.String()` that returns the output. I'm willing to defer that to followup work, though.

Comment (mheon) [reply libpod/image/tree.go]: (Honestly, though, I'd be fine keeping the tree local-only, and exposing only a literal output string over the API. There's precedence for this with things like `podman top`)

Comment (mheon) [reply libpod/image/tree.go]: (Or was it `podman stats`... I forget which)

**Pull request merged**